crates/services/executor/src/executor.rs

```diff
@@ -1030,228 +1080,671 @@ where
+                result: Some(state),
+                receipts,
+                total_gas: used_gas,
+                total_fee: tx_fee,
+            }
+        };
 
+        // queue up status for this tx to be stored once block id is finalized.
         execution_data.tx_status.push(TransactionExecutionStatus {
-            id: coinbase_id,
-            result: TransactionExecutionResult::Success {
-                result: None,
-                receipts: vec![],
-                total_gas: 0,
-                total_fee: 0,
-            },
+            id: tx_id,
+            result: status,
         });
+        Ok(())
+    }
 
-        if block_st_transaction
-            .storage::<ProcessedTransactions>()
-            .insert(&coinbase_id, &())?
-            .is_some()
+    fn update_input_used_gas<Tx>(
+        checked_tx: &mut Checked<Tx>,
+        tx_id: TxId,
+        tx: &mut Tx,
+    ) -> ExecutorResult<()>
+    where
+        Tx: ExecutableTransaction + PartialEq + Cacheable + Send + Sync + 'static,
+    {
+        for (original_input, produced_input) in checked_tx
+            .transaction()
+            .inputs()
+            .iter()
+            .zip(tx.inputs_mut())
         {
-            return Err(ExecutorError::TransactionIdCollision(coinbase_id))
+            let predicate_gas_used = original_input.predicate_gas_used();
+
+            if let Some(gas_used) = predicate_gas_used {
+                match produced_input {
+                    Input::CoinPredicate(CoinPredicate {
+                        predicate_gas_used, ..
+                    })
+                    | Input::MessageCoinPredicate(MessageCoinPredicate {
+                        predicate_gas_used,
+                        ..
+                    })
+                    | Input::MessageDataPredicate(MessageDataPredicate {
+                        predicate_gas_used,
+                        ..
+                    }) => {
+                        *predicate_gas_used = gas_used;
+                    }
+                    _ => {
+                        debug_assert!(false, "This error is not possible unless VM changes the order of inputs, \
+                        or we added a new predicate inputs.");
+                        return Err(ExecutorError::InvalidTransactionOutcome {
+                            transaction_id: tx_id,
+                        })
+                    }
+                }
+            }
         }
-        Ok(tx)
+        Ok(())
     }
 
-    #[allow(clippy::too_many_arguments)]
-    fn execute_chargeable_transaction<Tx, T>(
+    fn validate_utxos<Tx, T>(
         &self,
         mut checked_tx: Checked<Tx>,
         header: &PartialBlockHeader,
+        tx_st_transaction: &mut StorageTransaction<T>,
+    ) -> ExecutorResult<Checked<Tx>>
+    where
+        Tx: ExecutableTransaction + PartialEq + Cacheable + Send + Sync + 'static,
+        <Tx as IntoChecked>::Metadata: CheckedMetadata,
```

Comment:
```suggestion
        Tx: ExecutableTransaction + Cacheable,
```

You need to clean up constraints in a lot of places, please check contains for each new function that you've added=)

Reply:
I think it needs all of them except  the `PartialEq` on this one:
```rs
#[async_trait::async_trait]
impl<Tx> CheckPredicates for Checked<Tx>
where
    Tx: ExecutableTransaction + Send + Sync + 'static,
    <Tx as IntoChecked>::Metadata: crate::interpreter::CheckedMetadata + Send + Sync,
{
    fn check_predicates(
        mut self,
        params: &CheckPredicateParams,
    ) -> Result<Self, CheckError> {
        if !self.checks_bitmask.contains(Checks::Predicates) {
            Interpreter::<PredicateStorage, _>::check_predicates(&self, params)?;
            self.checks_bitmask.insert(Checks::Predicates);
        }
        Ok(self)
    }
```

Reply:
Okay, this function was a bad example=D But in other places we don't need so many constrains=)